From edc1e6033dae10f759b3c27a199c76f796729304 Mon Sep 17 00:00:00 2001
From: Evan Wang <xswang@marvell.com>
Date: Fri, 25 Aug 2017 18:12:19 +0800
Subject: [PATCH 1236/1345] fix: usb: xhci-plat: release USB phy after shut
 down

commit  6d19c5de742d01efa281921cc2cdc32d3293461b from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

During system reboot, the routine xhci_plat_remove() is called and the
USB PHY will be shut down twice. For the first time the USB PHY is
shut down directly by routine usb_phy_shutdown(), for second time it
is shut down within routine usb_remove_hcd() in case USB PHY is
associated(hcd->remove_phy will be valid). Since each time the USB
regulator is released during USB PHY shut down, the USB regulator will
be released twice and lead to warning.
In order to fix this issue, the USB PHY should be released after it is
shut down for the first time, thus during second time of shutdown, the
shutdown will be skipped since the USB PHY is already invalid.

Change-Id: Id74ecca489c1dc8665cc9d080ca990a1256be35d
Signed-off-by: Evan Wang <xswang@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/43388
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
(cherry picked from commit dc05dc5069c37ade909cb526738be6bc621a9db3)
Reviewed-on: http://vgitil04.il.marvell.com:8080/44222
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/host/xhci-plat.c |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/usb/host/xhci-plat.c b/drivers/usb/host/xhci-plat.c
index fcb9062..1f23d28 100644
--- a/drivers/usb/host/xhci-plat.c
+++ b/drivers/usb/host/xhci-plat.c
@@ -454,6 +454,8 @@ static int xhci_plat_remove(struct platform_device *dev)
 	} else {
 		usb_remove_hcd(xhci->shared_hcd);
 		usb_phy_shutdown(hcd->usb_phy);
+		usb_put_phy(hcd->usb_phy);
+		hcd->usb_phy = NULL;
 
 		usb_remove_hcd(hcd);
 	}
-- 
1.7.9.5

